File: 灵茶/图论/并查集/0797.py
from functools import cache
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)

        @cache
        def dfs(node):
            if node == n -1 :
                return[[n - 1]]

            ans = []
            for next in graph[node]:
                for res in dfs(next):
                    ans.append([node] + res)
            return ans
        
        return dfs(0)

File: 灵茶/图论/并查集/0797.java
import java.util.*;

class Solution {

    List<List<Integer>> ans = new ArrayList<>();

    List<Integer> path = new ArrayList<>();

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        path.add(0);
        dfs(graph, 0);
        return ans;
    }

    public void dfs(int[][] graph, int i) {
        if (i == graph.length - 1) {
            ans.add(new ArrayList<>(path));
            return;
        }
        for (int j : graph[i]) {
            path.add(j);
            dfs(graph, j);
            path.remove(path.size() - 1);
        }
    }
}